scripts-components/in-view/src/index.ts

```diff
@@ -0,0 +1,78 @@
+import { Component, ComponentInitializationError } from '@mangoweb/scripts-base'
+import { IsInView, isInView, isOutOfView, isSupported } from 'isinview'
+
+interface InViewData {
+	targets?: string
+	threshold?: number
+
+	isAboveViewClass?: string
+	isInViewClass?: string
+	isBelowViewClass?: string
+	isSeenClass?: string
+}
+
+export class InView extends Component<InViewData> {
+	static componentName = 'InView'
+
+	protected readonly targets: IsInView.Target
+	protected readonly options: Partial<IsInView.Options> = {}
+
+	protected readonly isAboveViewClass: string
+	protected readonly isInViewClass: string
+	protected readonly isBelowViewClass: string
+	protected readonly isSeenClass: string
+
+	public constructor(el: HTMLElement, data: InViewData) {
+		super(el, data)
+
+		if (!isSupported()) {
+			throw new ComponentInitializationError('InView component not supported. Add IntersectionObserver polyfill.')
+		}
+
+		this.targets = this.data.targets ? this.getChildren(this.data.targets) : this.el
+
+		if (this.data.threshold) {
+			this.options.threshold = this.data.threshold
+		}
+
+		this.isAboveViewClass = this.getProp('isAboveViewClass', 'is-aboveView')
+		this.isInViewClass = this.getProp('isInViewClass', 'is-inView')
+		this.isBelowViewClass = this.getProp('isBelowViewClass', 'is-belowView')
+		this.isSeenClass = this.getProp('isSeenClass', 'is-seen')
+	}
+
+	public init() {
+		isInView(
+			this.targets,
+			(target, details) => {
+				target.classList.add(this.isSeenClass)
+				this.togglePositionClasses(target, details)
+			},
+			this.options
+		)
+
+		isOutOfView(
+			this.targets,
+			(target, details) => {
+				this.togglePositionClasses(target, details)
+			},
+			this.options
+		)
+	}
```

Comment:
Co případ odebrání komponenty z DOMu? Budou na ní dále navěšeny event listeners této externí knihovny?

Reply:
Ta externí knihovna není nijak tajná, nemá ani obfuskovaný kód. 

Pro sledování elementů se používá `IntersectionObserver`. https://github.com/Onset/isinview/blob/master/src/index.ts#L48 Podobně jako v předchozí verzi `InView`. Listeners přežijí až do úplného zániku elementu. Takže elemenet můžeš odebrat z domu a zase ho vrátit a `InView` ho bude dále sledovat.

Každopádně díky za postrčení. Díky tobě jsem odhalil drobnou chybu, která logguje do konzole pži odebrání elementu z domu, když je na obrazovce. opravím.

Reply:
Ono zmiňované přežití listeneru může znamenat memory leak (např. v dynamicky generovaném seznamu), proto se ptám :)

Reply:
To je jasné.

Koukám do dokumentace a zdá se, že IntersectionObserver paměť neuvolní, když se na target ztratí reference.

Myslíš, že je potřeba to řešit? Leaky tam teoreticky nějaké být můžou, ale trackování odstranění elementu se ztrátou reference mi přijde v knihovně unreal. Mohl bych vracet třeba nějaký objekt/funkci, přes který/kterou se intersection observer dá vypnout, ale přijde mi to jako overkill.

`const destroy = isInView(...); destroy();`?

Reply:
@enzy, co myslíš?

Já bych kvůli jednoduchosti leakování zatím neřešil. (Nebylo řešeno ani v předchozí InView komponentě. AFAIK nám stejně leakuje celý komponentový systém, protože neuvolňuje reference, když se smaže place komponenta.)

Reply:
Nějaký draft `destroy()` metody pro Component mám na bici, poznamenávám do https://github.com/manGoweb/npm/issues/18